src/mpc-hc/SubtitlesProvider.h

```diff
@@ -0,0 +1,124 @@
+    ~P() { Uninitialize(); }                                                  \
+    P(P const&);                                                              \
+    P& operator=(P const&);                                                   \
+public:                                                                       \
+    static P& Instance() { static P that; return that; }                      \
+private:                                                                      \
+    using SubtitlesProvider::Login;                                           \
+    using SubtitlesProvider::Download;                                        \
+private:                                                                      \
+    virtual std::string Name() { return #P; }                                 \
+    virtual std::string Url() { return U; }                                   \
+    virtual std::string Languages();                                          \
+    virtual BOOL Flags(DWORD dwFlags) { return (dwFlags & (F)) == dwFlags; }  \
+    virtual int Icon() { return I; }                                          \
+private:                                                                      \
+    virtual SRESULT Search(const SubtitlesInfo& pFileInfo);                   \
+    virtual SRESULT Download(SubtitlesInfo& pSubtitlesInfo);                  \
+private:
+
+
+#define DEFINE_SUBTITLESPROVIDER_END                                          \
+};
+
+DEFINE_SUBTITLESPROVIDER_BEGIN(OpenSubtitles, "http://www.opensubtitles.org", IDI_OPENSUBTITLES, SPF_LOGIN | SPF_HASH | SPF_UPLOAD)
+virtual void Initialize();
+virtual void Uninitialize();
+virtual SRESULT Login(std::string& sUserName, std::string& sPassword);
+virtual SRESULT Hash(SubtitlesInfo& pFileInfo);
+virtual SRESULT Upload(const SubtitlesInfo& pSubtitlesInfo);
+XmlRpcClient* xmlrpc;
+XmlRpcValue token;
+DEFINE_SUBTITLESPROVIDER_END
+
+DEFINE_SUBTITLESPROVIDER_BEGIN(SubDB, "http://api.thesubdb.com", IDI_SUBDB, SPF_HASH | SPF_UPLOAD)
+virtual SRESULT Hash(SubtitlesInfo& pFileInfo);
+virtual SRESULT Upload(const SubtitlesInfo& pSubtitlesInfo);
+virtual std::string UserAgent() { return string_format("SubDB/1.0 (mpc-hc/%S; http://mpc-hc.org)", MPC_VERSION_STR); }
+DEFINE_SUBTITLESPROVIDER_END
+
+DEFINE_SUBTITLESPROVIDER_BEGIN(podnapisi, "http://www.podnapisi.net", IDI_PODNAPISI, SPF_SEARCH)
+virtual SRESULT Login(std::string& sUserName, std::string& sPassword);
+std::string GetLanguagesString();
+static const std::regex regex_pattern[];
+DEFINE_SUBTITLESPROVIDER_END
+
+DEFINE_SUBTITLESPROVIDER_BEGIN(titlovi, "http://www.titlovi.com", IDI_TITLOVI, SPF_SEARCH)
+std::string GetLanguagesString();
+DEFINE_SUBTITLESPROVIDER_END
+
+DEFINE_SUBTITLESPROVIDER_BEGIN(ysubs, "http://www.yifysubtitles.com", IDI_YSUBS, SPF_SEARCH)
+DEFINE_SUBTITLESPROVIDER_END
+
+
+static const struct { const char* code; const char* name; } podnapisi_languages[] = {
```

Comment:
Not all, but then again its a matter of performance trying to get them over the net.
In any case the list is expected to be extended on any change instead of rearranging items.